File: JS/src/add_two_numbers.js
import { ListNode } from "../lib/linked_list.js";

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @returns {ListNode}
 */
function addTwoNumbers(l1, l2) {
  let dummyHead = new ListNode();
  let tail = dummyHead;
  let carry = 0;

  while (l1 !== null || l2 !== null || carry !== 0) {
    let digit1 = l1?.val ?? 0;
    let digit2 = l2?.val ?? 0;

    let sum = digit1 + digit2 + carry;
    let digit = sum % 10;
    carry = Math.trunc(sum / 10);

    let newNode = new ListNode(digit, null);
    tail.next = newNode;
    tail = tail.next;

    l1 = l1?.next ?? null;
    l2 = l2?.next ?? null;
  }

  return dummyHead.next;
}

export { addTwoNumbers };

File: C/src/add_two_numbers.c
#include "../lib/linked_list.h"
#include <string.h>

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
  struct ListNode *res = NULL;
  int carry = 0;
  struct ListNode *current = res;

  while (l1 != NULL || l2 != NULL || carry > 0) {
    int digit1 = l1 != NULL ? l1->val : 0;
    int digit2 = l2 != NULL ? l2->val : 0;
    int sum = digit1 + digit2 + carry;

    int digit = sum % 10;
    carry = sum / 10;

    struct ListNode *new_node = malloc(sizeof(struct ListNode));
    new_node->val = digit;
    new_node->next = NULL;

    if (current == NULL) {
      current = new_node;
      res = current;
    } else {
      current->next = new_node;
      current = current->next;
    }

    l1 = l1 != NULL ? l1->next : NULL;
    l2 = l2 != NULL ? l2->next : NULL;
  }

  return res;
}
